File: src/SkyBlock/tiles/CatalystTile.php
<?php

namespace SkyBlock\tiles;

use pocketmine\block\Block;
use pocketmine\block\BlockTypeIds;
use pocketmine\block\tile\Nameable;
use pocketmine\block\tile\NameableTrait;
use pocketmine\block\tile\Spawnable;
use pocketmine\block\VanillaBlocks;
use pocketmine\math\Facing;
use pocketmine\nbt\tag\CompoundTag;
use SkyBlock\block\Catalyst;
use SkyBlock\Main;

class CatalystTile extends Spawnable implements Nameable {
    private int $delay = 200;

    use NameableTrait;

    public function onUpdate() : bool {
        if ($this->closed) return false;
        $block = $this->getBlock();
        if (!$block instanceof Catalyst) return true;
        if ($this->delay <= 0) {
            $this->delay = 200;
            if (($island = Main::getInstance()->getIslandManager()->getOnlineIslandByWorld($this->getPosition()->getWorld()->getDisplayName())) === null) return false;

            $up = $this->getPosition()->getWorld()->getBlock($block->getSide(Facing::UP)->getPosition());
            if ($up->getTypeId() === BlockTypeIds::AIR) {
                $this->getPosition()->getWorld()->setBlock($up->getPosition()->asVector3(), $this->getGenBlock(), false);
            }
        } else {
            $this->delay--;
        }
        return true;
    }

    private function getGenBlock() : Block {
        $island = Main::getInstance()->getIslandManager()->getOnlineIslandByWorld($this->getPosition()->getWorld()->getDisplayName());
        $oredata = $island->getOreDataPrefArray();

        $newBlock = $this->randomizeWithPercentage($oredata);
        $newBlock = match ($newBlock) {
            "coal" => VanillaBlocks::COAL_ORE(),
            "copper" => VanillaBlocks::COPPER_ORE(),
            "iron" => VanillaBlocks::IRON_ORE(),
            "lapis" => VanillaBlocks::LAPIS_LAZULI_ORE(),
            "gold" => VanillaBlocks::GOLD_ORE(),
            "diamond" => VanillaBlocks::DIAMOND_ORE(),
            "emerald" => VanillaBlocks::EMERALD_ORE(),
            "quartz" => VanillaBlocks::NETHER_QUARTZ_ORE(),
            "netherite" => VanillaBlocks::ANCIENT_DEBRIS(),
            "deep_coal" => VanillaBlocks::DEEPSLATE_COAL_ORE(),
            "deep_copper" => VanillaBlocks::DEEPSLATE_COPPER_ORE(),
            "deep_iron" => VanillaBlocks::DEEPSLATE_IRON_ORE(),
            "deep_lapis" => VanillaBlocks::DEEPSLATE_LAPIS_LAZULI_ORE(),
            "deep_gold" => VanillaBlocks::DEEPSLATE_GOLD_ORE(),
            "deep_diamond" => VanillaBlocks::DEEPSLATE_DIAMOND_ORE(),
            "deep_emerald" => VanillaBlocks::DEEPSLATE_EMERALD_ORE(),
            "deep_quartz" => VanillaBlocks::QUARTZ(),
            "deep_netherite" => VanillaBlocks::NETHERITE(),
            default => VanillaBlocks::COBBLESTONE(),
        };
        return $newBlock;
    }

    private function randomizeWithPercentage($array) : int|string {
        $totalWeight = array_sum($array);
        $rand = rand(1, $totalWeight);

        $currentWeight = 0;
        foreach ($array as $key => $weight) {
            $currentWeight += intval($weight);
            if ($rand <= $currentWeight) {
                return $key;
            }
        }
        return "default";
    }

    public function getName() : string {
        return "Catalyst";
    }

    public function getDefaultName() : string {
        return "Catalyst";
    }

    public function readSaveData(CompoundTag $nbt) : void {
    }

    protected function writeSaveData(CompoundTag $nbt) : void {
    }
}

File: SkyblockCore/src/main/java/dev/skyblock/tiles/CatalystTile.java
package dev.skyblock.tiles;

import dev.skyblock.SkyblockCore;
import dev.skyblock.island.Island;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

public class CatalystTile extends BaseTile {
    private static final int DELAY_TICKS = 200; // 10 seconds
    private int tickCounter = DELAY_TICKS;

    // Maps ore key names to Material — matches PHP source Catalyst ore list exactly
    private static final Map<String, Material> ORE_MAP = Map.ofEntries(
        Map.entry("coal",           Material.COAL_ORE),
        Map.entry("copper",         Material.COPPER_ORE),
        Map.entry("iron",           Material.IRON_ORE),
        Map.entry("lapis",          Material.LAPIS_ORE),
        Map.entry("gold",           Material.GOLD_ORE),
        Map.entry("diamond",        Material.DIAMOND_ORE),
        Map.entry("emerald",        Material.EMERALD_ORE),
        Map.entry("quartz",         Material.NETHER_QUARTZ_ORE),
        Map.entry("netherite",      Material.ANCIENT_DEBRIS),
        Map.entry("deep_coal",      Material.DEEPSLATE_COAL_ORE),
        Map.entry("deep_copper",    Material.DEEPSLATE_COPPER_ORE),
        Map.entry("deep_iron",      Material.DEEPSLATE_IRON_ORE),
        Map.entry("deep_lapis",     Material.DEEPSLATE_LAPIS_ORE),
        Map.entry("deep_gold",      Material.DEEPSLATE_GOLD_ORE),
        Map.entry("deep_diamond",   Material.DEEPSLATE_DIAMOND_ORE),
        Map.entry("deep_emerald",   Material.DEEPSLATE_EMERALD_ORE),
        Map.entry("deep_quartz",    Material.QUARTZ_BLOCK),
        Map.entry("deep_netherite", Material.NETHERITE_BLOCK)
    );

    public CatalystTile(Location location) {
        super(location);
    }

    @Override
    public void onUpdate() {
        if (tickCounter > 0) {
            tickCounter--;
            return;
        }
        tickCounter = DELAY_TICKS;

        SkyblockCore plugin = SkyblockCore.getInstance();
        String worldName = location.getWorld().getName();

        Island island = plugin.getIslandManager().getOnlineIslandByWorld(worldName).orElse(null);
        if (island == null) return;

        Block above = location.getBlock().getRelative(0, 1, 0);
        if (above.getType() != Material.AIR) return;

        Material oreToPlace = chooseOre(island.getOredatapref());
        above.setType(oreToPlace);
    }

    private Material chooseOre(Map<String, Integer> prefs) {
        if (prefs == null || prefs.isEmpty()) return Material.COBBLESTONE;

        int totalWeight = prefs.values().stream().mapToInt(Integer::intValue).sum();
        if (totalWeight <= 0) return Material.COBBLESTONE;

        int rand = ThreadLocalRandom.current().nextInt(1, totalWeight + 1);
        int cumulative = 0;
        for (Map.Entry<String, Integer> entry : prefs.entrySet()) {
            cumulative += entry.getValue();
            if (rand <= cumulative) {
                return ORE_MAP.getOrDefault(entry.getKey(), Material.COBBLESTONE);
            }
        }
        return Material.COBBLESTONE;
    }
}
